pkg/redisqueue/consumer.go

```diff
@@ -221,26 +221,46 @@ func (c *Consumer) work(ctx context.Context) {
 		}
 	}
 
+	// Check rate limit before we dequeue.
+	// Otherwise, we could dequeue the job but rate limited,
+	// and the server shutdowns, the job will be gone.
+	reservation := c.limiter.Reserve(c.limitBucket)
+	if reservation.Error() != nil {
+		c.logger.
+			WithField("tat", reservation.GetTimeToAct()).
+			WithField("bucket_key", c.limitBucket.Key()).
+			Debug("task rate limited")
+		select {
+		case <-c.shutdown:
+			return
+		case <-time.After(reservation.GetTimeToAct().Sub(c.clock.NowMonotonic())):
+			break
+		}
+	}
+
 	task, appProvider, err := c.dequeue(ctx)
 	if errors.Is(err, errNoTask) {
+		// There is actually no task.
+		// Cancel the reservation
+		c.logger.
+			WithField("tat", reservation.GetTimeToAct()).
+			WithField("bucket_key", c.limitBucket.Key()).
+			Debug("cancel reservation due to no task")
+		c.limiter.Cancel(reservation)
 		return
 	} else if err != nil {
 		c.logger.WithError(err).Error("failed to dequeue task")
 		c.dequeueBackoff.Increment()
 		return
 	}
 
-	c.dequeueBackoff.Reset()
+	c.logger.
+		WithField("tat", reservation.GetTimeToAct()).
+		WithField("bucket_key", c.limitBucket.Key()).
+		Debug("consume reservation")
 
-	if r := c.limiter.Reserve(c.limitBucket); r.Error() != nil {
```

Comment:
@louischan-oursky It's intentional to reserve after dequeue, otherwise every worker would eventually proceed to dequeue & run tasks simultaneously given enough time passed while queue is empty, due to the nature of rate limit token refill.